src/main/java/org/killbill/billing/plugin/dwolla/api/DwollaPaymentPluginApi.java

```diff
@@ -107,6 +114,57 @@ protected String getPaymentMethodId(DwollaPaymentMethodsRecord input) {
+        Map<String, String> mergedProperties = checkForDwollaDirectSolution(paymentMethodProps, properties, context);
+        final DateTime utcNow = clock.getUTCNow();
+        try {
+            dao.addPaymentMethod(kbAccountId, kbPaymentMethodId, setDefault, mergedProperties, utcNow, context.getTenantId());
+        } catch (final SQLException e) {
+            throw new PaymentPluginApiException("Unable to add payment method for kbPaymentMethodId " + kbPaymentMethodId, e);
+        }
+    }
+
+    private Map<String, String> checkForDwollaDirectSolution(final PaymentMethodPlugin paymentMethodProps, final Iterable<PluginProperty> properties, final CallContext context) throws PaymentPluginApiException {
+        final Map<String, String> mergedProperties = new HashMap<String, String>(PluginProperties.toStringMap(paymentMethodProps.getProperties(), properties));
+        final String code = mergedProperties.get(CODE);
+        if (!Strings.isNullOrEmpty(code)) {
+            // Dwolla Direct - needs to retrieve account id and funding sources
+
+            final HalLink dwollaAccountLink;
+            final TokenResponse tokenResponse = client.getUserToken(code);
+            try {
+                client.getClient().setAccessToken(tokenResponse.access_token);
+                final CatalogResponse root = client.getRootInfo();
+                dwollaAccountLink = root.getLinks().get(ACCOUNT);
+            } catch (ApiException e) {
+                throw new PaymentPluginApiException("Unable to get account from Dwolla Direct auth code: " + code, e);
+            }
+
+            final String dwollaAccountId = getIdFromUrl(dwollaAccountLink);
+            try {
+                // save/update Dwolla Direct account token pair
+                if (dao.getTokens(dwollaAccountId, context.getTenantId()) == null) {
+                    dao.addTokens(tokenResponse.access_token, tokenResponse.refresh_token, dwollaAccountId, context.getTenantId());
+                } else {
+                    dao.updateTokens(tokenResponse.access_token, tokenResponse.refresh_token, dwollaAccountId, context.getTenantId());
+                }
+            } catch (SQLException e) {
+                throw new PaymentPluginApiException("Error saving Dwolla Direct account token pair for " + dwollaAccountLink.getHref(), e);
+            }
+
+            final HalLink fundingSourceLink = getFirstActiveFundingSource(dwollaAccountLink.getHref());
+            final String fundingSourceId = getIdFromUrl(fundingSourceLink);
+            mergedProperties.put(PROPERTY_ACCOUNT_ID, dwollaAccountId);
+            mergedProperties.put(PROPERTY_FUNDING_SOURCE_ID, fundingSourceId);
+        }
+        return mergedProperties;
+    }
+
+    private String getIdFromUrl(HalLink halLink) {
+        return halLink.getHref().substring(halLink.getHref().lastIndexOf(47) + 1);
```

Comment:
Dwolla's responses always contains the full url and we need to isolate the id. I took that idea from [their client](https://github.com/Dwolla/dwolla-swagger-java/blob/master/src/main/java/io/swagger/client/api/FundingsourcesApi.java#L76).